File: back/HistoryBank/src/main/java/com/majortomdev/historybank/service/NoteService.java
package com.majortomdev.historybank.service;

import com.majortomdev.historybank.entities.Note;
import com.majortomdev.historybank.entities.User;
import com.majortomdev.historybank.repository.NoteRepository;
import com.majortomdev.historybank.repository.UserRepository;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Optional;

@Service
public class NoteService {

    private final NoteRepository noteRepository;
    private final UserRepository userRepository;

    public NoteService(NoteRepository noteRepository, UserRepository userRepository) {
        this.noteRepository = noteRepository;
        this.userRepository = userRepository;
    }

    public Note createNote(Note note, String username) {
        User user = userRepository.findByUsername(username);
        if (user == null) {
            throw new RuntimeException("User not found");
        }

        Note newNote = new Note(note.getTitle(), note.getNotes(), user);
        return noteRepository.save(newNote);
    }

    public List<Note> getNotesForUser(String username) {
        User user = userRepository.findByUsername(username);
        if (user == null) {
            throw new RuntimeException("User not found");
        }

        return noteRepository.findByUserId(user.getId());
    }

    public Note updateNote(int noteId, Note noteDetails, String username) {
        User user = userRepository.findByUsername(username);
        if (user == null) {
            throw new RuntimeException("User not found");
        }

        Note note = noteRepository.findById(noteId)
                .orElseThrow(() -> new RuntimeException("Note not found"));

        if (note.getUser().getId() != user.getId()) {
            throw new RuntimeException("You are not authorized to update this note");
        }

        note.setTitle(noteDetails.getTitle());
        note.setNotes(noteDetails.getNotes());
        return noteRepository.save(note);
    }

    public void deleteNote(int noteId, String username) {
        User user = userRepository.findByUsername(username);
        if (user == null) {
            throw new RuntimeException("User not found");
        }

        Note note = noteRepository.findById(noteId)
                .orElseThrow(() -> new RuntimeException("Note not found"));

        if (note.getUser().getId() != user.getId()) {
            throw new RuntimeException("You are not authorized to delete this note");
        }

        noteRepository.delete(note);
    }
}

File: back/HistoryBank/src/main/kotlin/com/majortomdev/HistoryBank/Service/NoteService.kt
@Service
class NoteService(
    private val noteRepository: NoteRepository,
    private val userRepository: UserRepository
) {

    fun createNote(note: Note, username: String): Note {
        val user = userRepository.findByUsername(username)
            ?: throw RuntimeException("User not found")

        val newNote = Note(title = note.title, notes = note.notes, user = user)
        return noteRepository.save(newNote)
    }

    fun getNotesForUser(username: String): List<Note> {
        val user = userRepository.findByUsername(username)
            ?: throw RuntimeException("User not found")

        return noteRepository.findByUserId(user.id)
    }

    fun updateNote(noteId: Int, noteDetails: Note, username: String): Note {
        val user = userRepository.findByUsername(username)
            ?: throw RuntimeException("User not found")

        val note = noteRepository.findById(noteId)
            .orElseThrow { RuntimeException("Note not found") }

        if (note.user.id != user.id) {
            throw RuntimeException("You are not authorized to update this note")
        }

        note.title = noteDetails.title
        note.notes = noteDetails.notes
        return noteRepository.save(note)
    }

    fun deleteNote(noteId: Int, username: String) {
        val user = userRepository.findByUsername(username)
            ?: throw RuntimeException("User not found")

        val note = noteRepository.findById(noteId)
            .orElseThrow { RuntimeException("Note not found") }

        if (note.user.id != user.id) {
            throw RuntimeException("You are not authorized to delete this note")
        }

        noteRepository.delete(note)
    }
}
